### PR TITLE
Feature/quantile series

### LINKED ISSUE #219: Histogram measures
Histograms should be special measures. 

- [ ] ~~In config they should be defined as such. No formula, but field for type, reference to underlying measure and default percentile.~~
- [x] In UI user should be able to define what percentile he likes to calculate. At start is set to default value.
- [x] It is possible to add multiple histograms to Series. If you try to add second histogram with the same percentile, "OK" action should be disabled thus user need to provide different percentile to apply measure.
- [x] Tune parameters (resolution, numBuckets, lowerLimit, upperLimit) are taken from Data Cube config

### PR DESCRIPTION
**Pull request by adrianmroz**
Closes #219

### CODE REVIEW

File: src/client/components/series-menu/quantile-picker.tsx [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import * as React from "react";
+import { Omit } from "../../../common/utils/functional/functional";
+import { InputWithPresets, InputWithPresetsProps } from "../input-with-presets/input-with-presets";
+
+type NumberInputWithPresetsProps = Omit<InputWithPresetsProps<number>, "parseCustomValue" | "formatCustomValue">;
```
Comment (adrianmroz) [inline src/client/components/series-menu/quantile-picker.tsx]: Rename props type

File: src/client/components/measure-actions-menu/add-quantile-series.tsx [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+  appendDirtySeries: Unary<Series, void>;
+  measure: Measure;
+  series: SeriesList;
+  onClose: Fn;
+}
+
+export const AddQuantileSeriesButton: React.SFC<AddQuantileSeriesButtonProps> = props => {
+  const { series, measure, appendDirtySeries, addSeries, onClose } = props;
+
+  const disabled = !measure.isQuantile();
```
Comment (adrianmroz) [inline src/client/components/measure-actions-menu/add-quantile-series.tsx]: No disabled is needed

File: src/client/components/measure-actions-menu/add-quantile-series.tsx [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+
+  function onNewQuantileSeries() {
+    if (!disabled) {
```
Comment (adrianmroz) [inline src/client/components/measure-actions-menu/add-quantile-series.tsx]: Throw if measure doesn't have quantile expression. Maybe do that in `fromQuantileMeasure`

File: src/common/models/series/series.ts [hunk 4]
```diff
@@ -29,15 +32,31 @@ export interface SeriesBehaviours {
...
 
-export function fromJS(params: any): Series {
+export function fromMeasure(measure: Measure): MeasureSeries | QuantileSeries {
+  if (measure.expression instanceof QuantileExpression) {
+    return QuantileSeries.fromQuantileMeasure(measure);
+  }
+  return MeasureSeries.fromMeasure(measure);
+}
+
+function fromJSWithMeasure(params: any, { expression }: Measure): MeasureSeries | QuantileSeries {
```
Comment (adrianmroz) [inline src/common/models/series/series.ts]: Better name

File: src/client/components/input-with-presets/input-with-presets.tsx [hunk 5]
```diff
@@ -20,67 +20,72 @@ import { classNames } from "../../utils/dom/dom";
...
-export class InputWithPresets extends React.Component<InputWithPresetsProps, InputWithPresetsState> {
+export class InputWithPresets<T = string> extends React.Component<InputWithPresetsProps<T>, InputWithPresetsState> {
 
   initialState(): InputWithPresetsState {
-    const { selected, presets } = this.props;
+    const { selected, presets, formatCustomValue } = this.props;
     const presetPicked = presets.some(({ identity }) => identity === selected);
     const customPicked = selected !== undefined && !presetPicked;
-    const customValue = presetPicked ? "" : (selected || "");
+    const customValue = presetPicked ? "" : (selected ? formatCustomValue(selected) : "");
```
Comment (erykpiast) [inline src/client/components/input-with-presets/input-with-presets.tsx]: extract it to the function, please.

File: src/common/models/series/quantile-concrete-series.ts [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+
+import { ApplyExpression, Expression as PlywoodExpression, QuantileExpression } from "plywood";
+import { Measure } from "../measure/measure";
+import { ConcreteSeries, SeriesDerivation } from "./concrete-series";
+import { QuantileSeries } from "./quantile-series";
+
+export class QuantileConcreteSeries extends ConcreteSeries<QuantileSeries> {
+
+  constructor(series: QuantileSeries, measure: Measure) {
+    super(series, measure);
```
Comment (erykpiast) [inline src/common/models/series/quantile-concrete-series.ts]: You don't really need this.

File: src/common/models/series/quantile-series.ts [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+  static fromQuantileMeasure({ name: reference, expression }: Measure) {
+    if (!(expression instanceof QuantileExpression)) throw new Error(`Expected QuantileExpression, got ${expression}`);
+    return new QuantileSeries({
+      reference,
+      percentile: expression.value * 100
+    });
+  }
+
+  constructor(params: RequireOnly<QuantileSeriesValue, "percentile" | "reference">) {
+    super(params);
```
Comment (erykpiast) [inline src/common/models/series/quantile-series.ts]: Check this place as well.

File: src/common/models/series/series.ts (hunk 4 above)
Comment (erykpiast) [inline src/common/models/series/series.ts]: `unknown`. !!!

File: src/common/models/series/series.ts [hunk 8]
```diff
@@ -29,15 +32,31 @@ export interface SeriesBehaviours {
...
+  if (expression instanceof QuantileExpression) {
+    return QuantileSeries.fromJS(params);
+  }
+  return MeasureSeries.fromJS(params);
+}
+
+export function fromJS(params: any, measure: Measure): Series {
```
Comment (erykpiast) [inline src/common/models/series/series.ts]: !!!

File: src/client/components/measure-actions-menu/add-quantile-series.tsx (hunk 2 above)
Comment (erykpiast) [reply src/client/components/measure-actions-menu/add-quantile-series.tsx]: Work on your grammar, boy.

File: src/client/components/series-menu/quantile-picker.tsx [hunk 9]
```diff
@@ -0,0 +1,33 @@
...
+
+function parse(s: string): number | null {
```
Comment (erykpiast) [inline src/client/components/series-menu/quantile-picker.tsx]: It's always a `number`, dude.

File: src/client/components/series-menu/quantile-series-menu.tsx [hunk 10]
```diff
@@ -0,0 +1,88 @@
...
+  { identity: 99, name: "99" }
+];
+
+export const QuantileSeriesMenu: React.SFC<QuantileSeriesMenuProps> = ({ seriesList, initialSeries, measure, series, onChange }) => {
+
+  function isSeriesValid(series: QuantileSeries) {
+    return validateSeries(series) === null;
+  }
+
+  function validateSeries(series: QuantileSeries): string | null {
```
Comment (erykpiast) [inline src/client/components/series-menu/quantile-series-menu.tsx]: Do not rely on hoisting, please. Move this function above `isSeriesValid`. Please.

File: src/client/components/series-tile/series-tiles-row.tsx [hunk 11]
```diff
@@ -140,28 +150,52 @@ export class SeriesTilesRow extends React.Component<SeriesTilesRowProps, SeriesT
...
+    } else {
+      this.dropNewSeries(fromMeasure(DragManager.draggingMeasure()), this.calculateDragPosition(e));
+    }
+  }
 
-    if (dragPosition.replace === this.maxItems()) {
-      dragPosition = new DragPosition({ insert: dragPosition.replace });
+  private dropNewSeries(newSeries: Series, dragPosition: DragPosition) {
+    const { clicker, essence: { series } } = this.props;
+    const isSeriesValid = newSeries instanceof MeasureSeries || (newSeries instanceof QuantileSeries && !series.hasSeries(newSeries));
```
Comment (erykpiast) [inline src/client/components/series-tile/series-tiles-row.tsx]: Change the condition to `if (isDuplicated)`, invert `if` and `else` blocks. It's gonna be more readable, hopefully.

File: src/client/components/series-tile/series-tiles-row.tsx [hunk 12]
```diff
@@ -140,28 +150,52 @@ export class SeriesTilesRow extends React.Component<SeriesTilesRowProps, SeriesT
...
   }
 
   appendMeasureSeries = (measure: Measure) => {
-    this.props.clicker.addSeries(MeasureSeries.fromMeasure(measure));
+    const series = fromMeasure(measure);
+    if (series instanceof MeasureSeries) {
+      this.props.clicker.addSeries(series);
+      return;
+    }
+    if (!this.props.essence.series.hasSeries(series)) {
```
Comment (erykpiast) [inline src/client/components/series-tile/series-tiles-row.tsx]: merge these two `if`s with `||` and name partial results.

File: src/client/views/cube-view/cube-view.tsx [hunk 13]
```diff
@@ -474,9 +474,9 @@ export class CubeView extends React.Component<CubeViewProps, CubeViewState> {
     (this.refs["filterTile"] as FilterTile).filterMenuRequest(dimension);
   }
 
-  newExpressionSeries = (series: Series) => {
+  appendDirtySeries = (series: Series) => {
     if (!series) return;
-    (this.refs["seriesTile"] as SeriesTilesRow).newExpressionSeries(series);
+    (this.refs["seriesTile"] as SeriesTilesRow).appendDirtySeries(series);
```
Comment (erykpiast) [inline src/client/views/cube-view/cube-view.tsx]: Dot notation, please.

File: src/common/models/series/series.ts [hunk 14]
```diff
@@ -43,9 +43,11 @@ export function fromMeasure(measure: Measure): MeasureSeries | QuantileSeries {
 
 function inferTypeAndConstruct({ expression }: Measure, params: any): MeasureSeries | QuantileSeries {
   if (expression instanceof QuantileExpression) {
-    return QuantileSeries.fromJS(params);
+    const type = SeriesType.QUANTILE;
```
Comment (erykpiast) [inline src/common/models/series/series.ts]: This `const` gives you nothing. Remove it, please.

File: src/common/models/series/quantile-series.mocha.ts [hunk 15]
```diff
@@ -0,0 +1,68 @@
...
+  name: "my-quantile",
+  formula: "$main.quantile($histogram, 0.92, 'tuning')"
+});
+
+const quantileSeries = QuantileSeries.fromQuantileMeasure(quantileMeasure);
+
+describe("QuantileSeries", () => {
+  describe("fromQuantileMeasure", () => {
+    it("throws when measure expression is not a quantile", () => {
+      expect(() => QuantileSeries.fromQuantileMeasure(MeasureFixtures.wikiCount())).throws();
```
Comment (erykpiast) [inline src/common/models/series/quantile-series.mocha.ts]: You can specify the expected message pattern, like this: `throws(/foo bar/)`

File: src/common/models/series/series.ts (hunk 4 above)
Comment (adrianmroz) [reply src/common/models/series/series.ts]: It is edge of the system. Somewhere you need any. Or unknown with full validation, which would be nice but out of scope for this task.

File: src/common/models/series/series.ts (hunk 8 above)
Comment (adrianmroz) [reply src/common/models/series/series.ts]: ditto

**Pull request merged**